git repository upgrade

Byron opened the pull request:
This definitely shows some issues with the current exposure of imara-diff.

This should have been covered by `git-diff`, but here we are.

Byron committed:
upgrade to `git-repoitory v0.26`, fail to make diff compile.

This will work, just needs more massaging.

Byron committed:
try to rewrite explicit trait impl as closure, same lifeimte issues.

This seems like a systematic issue with `lines()` that needs fixing.

pascalkuthe commented on src/index/diff/delegate.rs:
```diff
@@ -55,25 +56,56 @@ impl Delegate {
-                        }
+                        },
+                    );
+                    for op in line_changes.drain(..) {
+                        let (version, op) = op?;
+                        match op {
+                            Op::Add => {
+                                self.changes.push(if version.yanked {
+                                    Change::Yanked(version)
+                                } else {
+                                    Change::Added(version)
+                                });
+                            }
+                            Op::Delete => {
+                                self.delete_version_ids.insert(version.id());
+                            }
+                        };
```
Why are you collecting into a vec that you then later iterate on again, instead of doing this in the closure directly? I don't see any reasons that the changes would need to be in a vec first. Would save an allocation and maybe make the code simpler this also avoids the `Op` enum

pascalkuthe replied:
~~This shoulld also fix the lifetime error you are getting here~~

Byron replied:
Thanks for having taken a look!

I think the callback isn't allowed to return anything, and the results have to be kept. Generally I am happy with `Vec` as long as the allocation can be reused. Here done via `push()` and `drain(..)`.

I am now working on making this case easy to handle using the `lines()` method.

Byron committed:
A first stab at getting the diff back to work with `imara-diff`

There are a lot of test failures, hopefully an easy fix due to something
fundamental being wrong.

Byron committed:
fix most of the tests

Byron committed:
one more test bytes the dust, but…

…it looks like normalization is now broken, probably some incorrect
handling of modifications, which previously only showed up as deletion
and addition.

Byron committed:
all tests are green

Byron committed:
upgrade to git-repository v0.27
